File: game_logic.py
import networkx as nx
from typing import Dict, List, Set, Tuple, Any, Optional
import re

class PackageDependencyGame:
    """Main game logic for package dependency resolution"""
    
    def __init__(self, dependency_graph: nx.DiGraph, root_package: str):
        self.dependency_graph = dependency_graph.copy()
        self.root_package = root_package
        self.selected_packages: Set[str] = set()
        
        # Validate inputs
        if root_package not in dependency_graph.nodes():
            raise ValueError(f"Root package {root_package} not found in dependency graph")
    
    def parse_package_node(self, node: str) -> Tuple[str, str]:
        """Parse package node to extract package name and version"""
        # Assuming format like "package_name-1.0.0" or "package_name==1.0.0"
        if '-' in node:
            parts = node.rsplit('-', 1)
            if len(parts) == 2 and '.' in parts[1]:
                return parts[0], parts[1]
        elif '==' in node:
            parts = node.split('==')
            if len(parts) == 2:
                return parts[0], parts[1]
        
        # Fallback: treat as package name with default version
        return node, "1.0.0"
    
    def get_package_name(self, node: str) -> str:
        """Get package name without version"""
        return self.parse_package_node(node)[0]
    
    def get_package_version(self, node: str) -> str:
        """Get package version"""
        return self.parse_package_node(node)[1]
    
    def get_hierarchical_layout(self) -> Dict[str, Tuple[float, float]]:
        """Create hierarchical layout with packages organized by dependency levels"""
        # Calculate levels using topological sort
        try:
            # Check if graph is acyclic first
            if not nx.is_directed_acyclic_graph(self.dependency_graph):
                # If it has cycles, use spring layout
                return nx.spring_layout(self.dependency_graph, k=2, iterations=50)
            
            levels = {}
            topo_order = list(nx.topological_sort(self.dependency_graph))
            
            # Assign levels based on longest path from root
            for node in topo_order:
                if self.dependency_graph.in_degree(node) == 0:
                    levels[node] = 0
                else:
                    max_pred_level = -1
                    for pred in self.dependency_graph.predecessors(node):
                        if pred in levels:
                            max_pred_level = max(max_pred_level, levels[pred])
                    levels[node] = max_pred_level + 1
            
            # Group packages by level and by package name
            level_groups = {}
            for node, level in levels.items():
                if level not in level_groups:
                    level_groups[level] = {}
                
                package_name = self.get_package_name(node)
                if package_name not in level_groups[level]:
                    level_groups[level][package_name] = []
                level_groups[level][package_name].append(node)
            
            # Calculate positions
            pos = {}
            y_spacing = 2.0
            x_spacing = 2.0
            
            for level, package_groups in level_groups.items():
                y = -level * y_spacing  # Higher levels at top
                
                total_packages = sum(len(versions) for versions in package_groups.values())
                start_x = -(total_packages - 1) * x_spacing / 2
                
                x_offset = 0
                for package_name, versions in sorted(package_groups.items()):
                    for i, version_node in enumerate(sorted(versions)):
                        x = start_x + x_offset * x_spacing
                        pos[version_node] = (x, y + i * 0.3)  # Slight vertical offset for versions
                        x_offset += 1
            
            return pos
            
        except nx.NetworkXError:
            # Fallback to spring layout if graph has cycles
            return nx.spring_layout(self.dependency_graph, k=2, iterations=50)
    
    def select_package(self, package: str) -> bool:
        """Select a package. Returns True if selection was successful."""
        if package not in self.dependency_graph.nodes():
            return False
        
        # Check if selecting this package would violate version constraints
        package_name = self.get_package_name(package)
        for selected in self.selected_packages:
            if self.get_package_name(selected) == package_name:
                # Cannot select two versions of the same package
                return False
        
        self.selected_packages.add(package)
        return True
    
    def deselect_package(self, package: str) -> bool:
        """Deselect a package. Returns True if deselection was successful."""
        if package in self.selected_packages:
            self.selected_packages.remove(package)
            return True
        return False
    
    def check_constraints(self) -> List[str]:
        """Check all constraints and return list of violations"""
        violations = []
        
        # Check for multiple versions of same package
        package_names = {}
        for package in self.selected_packages:
            name = self.get_package_name(package)
            if name not in package_names:
                package_names[name] = []
            package_names[name].append(package)
        
        for name, versions in package_names.items():
            if len(versions) > 1:
                violations.append(f"Multiple versions selected for {name}: {', '.join(versions)}")
        
        # Check dependency satisfaction
        for package in self.selected_packages:
            dependencies = list(self.dependency_graph.successors(package))
            for dep in dependencies:
                dep_name = self.get_package_name(dep)
                
                # Check if any version of this dependency is selected
                dep_satisfied = False
                for selected in self.selected_packages:
                    if self.get_package_name(selected) == dep_name:
                        dep_satisfied = True
                        break
                
                if not dep_satisfied:
                    violations.append(f"{package} requires {dep_name} but it's not selected")
        
        return violations
    
    def is_valid_solution(self) -> bool:
        """Check if current selection is a valid solution using boolean solver"""
        from boolean_solver import BooleanSolver
        
        # Use boolean solver to check if all clauses are satisfied
        boolean_solver = BooleanSolver(self.dependency_graph, self.root_package)
        all_satisfied, _ = boolean_solver.evaluate_all_clauses(self.selected_packages)
        return all_satisfied
    
    def get_installable_packages(self) -> Set[str]:
        """Get set of packages that can be successfully installed with current selection"""
        installable = set()
        
        def can_install(package: str, visited: Optional[Set[str]] = None) -> bool:
            if visited is None:
                visited = set()
            assert visited is not None  # Type hint for mypy
            
            if package in visited:
                return True  # Assume already handled
            
            visited.add(package)
            
            # Check if package is selected
            package_name = self.get_package_name(package)
            package_selected = False
            for selected in self.selected_packages:
                if self.get_package_name(selected) == package_name:
                    package_selected = True
                    break
            
            if not package_selected:
                return False
            
            # Check dependencies
            dependencies = list(self.dependency_graph.successors(package))
            for dep in dependencies:
                if not can_install(dep, visited.copy()):
                    return False
            
            return True
        
        for package in self.selected_packages:
            if can_install(package):
                installable.add(package)
        
        return installable
    
    def get_game_state(self) -> Dict[str, Any]:
        """Get current game state for serialization/debugging"""
        return {
            'root_package': self.root_package,
            'selected_packages': list(self.selected_packages),
            'constraint_violations': self.check_constraints(),
            'is_valid_solution': self.is_valid_solution(),
            'total_packages': self.dependency_graph.number_of_nodes(),
            'total_dependencies': self.dependency_graph.number_of_edges()
        }

File: docs/game-logic.js
// Package Dependency Game Logic - JavaScript port of the Python game_logic.py

class PackageDependencyGame {
    constructor(dependencyGraph, rootPackage) {
        this.dependencyGraph = dependencyGraph;
        this.rootPackage = rootPackage;
        this.selectedPackages = new Set();
        
        // Validate inputs
        if (!dependencyGraph.nodes.includes(rootPackage)) {
            throw new Error(`Root package ${rootPackage} not found in dependency graph`);
        }
    }

    parsePackageNode(node) {
        // Parse package node to extract package name and version
        if (node.includes('-') && node.split('-').length === 2) {
            const parts = node.split('-');
            if (parts.length === 2 && parts[1].includes('.')) {
                return [parts[0], parts[1]];
            }
        } else if (node.includes('==')) {
            const parts = node.split('==');
            if (parts.length === 2) {
                return [parts[0], parts[1]];
            }
        }
        
        // Fallback: treat as package name with default version
        return [node, "1.0.0"];
    }

    getPackageName(node) {
        return this.parsePackageNode(node)[0];
    }

    getPackageVersion(node) {
        return this.parsePackageNode(node)[1];
    }

    getHierarchicalLayout() {
        // Create hierarchical layout for the graph visualization
        const levels = new Map();
        const visited = new Set();
        const inDegree = new Map();
        
        // Calculate in-degrees
        for (const node of this.dependencyGraph.nodes) {
            inDegree.set(node, 0);
        }
        
        for (const [source, target] of this.dependencyGraph.edges) {
            inDegree.set(target, inDegree.get(target) + 1);
        }
        
        // Topological sort to assign levels
        const queue = [];
        for (const [node, degree] of inDegree.entries()) {
            if (degree === 0) {
                levels.set(node, 0);
                queue.push(node);
            }
        }
        
        while (queue.length > 0) {
            const current = queue.shift();
            const currentLevel = levels.get(current);
            
            // Find all nodes that depend on current
            for (const [source, target] of this.dependencyGraph.edges) {
                if (source === current) {
                    const newDegree = inDegree.get(target) - 1;
                    inDegree.set(target, newDegree);
                    
                    if (newDegree === 0) {
                        levels.set(target, currentLevel + 1);
                        queue.push(target);
                    }
                }
            }
        }
        
        // Group packages by level and by package name
        const levelGroups = new Map();
        for (const [node, level] of levels.entries()) {
            if (!levelGroups.has(level)) {
                levelGroups.set(level, new Map());
            }
            
            const packageName = this.getPackageName(node);
            if (!levelGroups.get(level).has(packageName)) {
                levelGroups.get(level).set(packageName, []);
            }
            levelGroups.get(level).get(packageName).push(node);
        }
        
        // Calculate positions
        const positions = new Map();
        const ySpacing = 120;
        const xSpacing = 150;
        
        for (const [level, packageGroups] of levelGroups.entries()) {
            const y = level * ySpacing;
            
            let totalPackages = 0;
            for (const versions of packageGroups.values()) {
                totalPackages += versions.length;
            }
            
            const startX = -(totalPackages - 1) * xSpacing / 2;
            let xOffset = 0;
            
            for (const [packageName, versions] of Array.from(packageGroups.entries()).sort()) {
                for (let i = 0; i < versions.length; i++) {
                    const versionNode = versions[i];
                    const x = startX + xOffset * xSpacing;
                    positions.set(versionNode, [x, y + i * 30]); // Slight vertical offset for versions
                    xOffset++;
                }
            }
        }
        
        return positions;
    }

    selectPackage(packageNode) {
        // Select a package. Returns 1 if selection was successful, 0 if version conflict, -1 if error
        if (!this.dependencyGraph.nodes.includes(packageNode)) {
            return -1;
        }
        
        // Check if selecting this package would violate version constraints
        const packageName = this.getPackageName(packageNode);
        for (const selected of this.selectedPackages) {
            if (this.getPackageName(selected) === packageName) {
                // Cannot select two versions of the same package
                return 0;
            }
        }
        
        this.selectedPackages.add(packageNode);
        return 1;
    }

    deselectPackage(packageNode) {
        // Deselect a package. Returns true if deselection was successful
        if (this.selectedPackages.has(packageNode)) {
            this.selectedPackages.delete(packageNode);
            return true;
        }
        return false;
    }

    checkConstraints() {
        // Check all constraints and return list of violations
        const violations = [];
        
        // Check for multiple versions of same package
        const packageNames = new Map();
        for (const pkg of this.selectedPackages) {
            const name = this.getPackageName(pkg);
            if (!packageNames.has(name)) {
                packageNames.set(name, []);
            }
            packageNames.get(name).push(pkg);
        }
        
        for (const [name, versions] of packageNames.entries()) {
            if (versions.length > 1) {
                violations.push(`Multiple versions selected for ${name}: ${versions.join(', ')}`);
            }
        }
        
        // Check dependency satisfaction
        for (const pkg of this.selectedPackages) {
            const dependencies = this.getDependencies(pkg);
            for (const dep of dependencies) {
                const depName = this.getPackageName(dep);
                
                // Check if any version of this dependency is selected
                let depSatisfied = false;
                for (const selected of this.selectedPackages) {
                    if (this.getPackageName(selected) === depName) {
                        depSatisfied = true;
                        break;
                    }
                }
                
                if (!depSatisfied) {
                    violations.push(`${pkg} requires ${depName} but it's not selected`);
                }
            }
        }
        
        return violations;
    }

    getDependencies(packageNode) {
        // Get direct dependencies of a package
        const dependencies = [];
        for (const [source, target] of this.dependencyGraph.edges) {
            if (source === packageNode) {
                dependencies.push(target);
            }
        }
        return dependencies;
    }

    isValidSolution() {
        // Check if current selection is a valid solution using boolean solver
        try {
            const booleanSolver = new BooleanSolver(this.dependencyGraph, this.rootPackage);
            const [allSatisfied, _] = booleanSolver.evaluateAllClauses(this.selectedPackages);
            return allSatisfied;
        } catch (error) {
            console.error('Error checking solution validity:', error);
            return false;
        }
    }

    getGameState() {
        // Get current game state for serialization/debugging
        return {
            rootPackage: this.rootPackage,
            selectedPackages: Array.from(this.selectedPackages),
            constraintViolations: this.checkConstraints(),
            isValidSolution: this.isValidSolution(),
            totalPackages: this.dependencyGraph.nodes.length,
            totalDependencies: this.dependencyGraph.edges.length
        };
    }

    reset() {
        // Reset the game state
        this.selectedPackages.clear();
    }
}

// Export for use in other files
if (typeof module !== 'undefined' && module.exports) {
    module.exports = PackageDependencyGame;
}
